### PR TITLE
Add support for static provisioning of file systems and child volumes

### PR DESCRIPTION
**Pull request by hughdanliu**
Adds support for static provisioning of FSx for OpenZFS file systems and child volumes as persistent volumes in Kubernetes clusters. 

**TESTING**:
Successful mount of an FSx for OpenZFS file system (mounts the root volume):
```
[hughliu@dev-dsk-hughliu-1e-2e6bb5a9 aws-fsx-openzfs-csi-driver]$ k apply -f examples/static_provisioning/file_systems/
pod/fsx-openzfs-app created
persistentvolume/fsx-openzfs-pv created
persistentvolumeclaim/fsx-openzfs-claim created
[hughliu@dev-dsk-hughliu-1e-2e6bb5a9 aws-fsx-openzfs-csi-driver]$ k describe pv fsx-openzfs-pv
Name:            fsx-openzfs-pv
Labels:          <none>
Annotations:     pv.kubernetes.io/bound-by-controller: yes
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    
Status:          Bound
Claim:           default/fsx-openzfs-claim
Reclaim Policy:  Delete
Access Modes:    RWX
VolumeMode:      Filesystem
Capacity:        100Gi
Node Affinity:   <none>
Message:         
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            fsx.openzfs.csi.aws.com
    FSType:            
    VolumeHandle:      fs-0edcf19e7e13f53ff
    ReadOnly:          false
    VolumeAttributes:      dnsname=fs-0edcf19e7e13f53ff.fsx.us-east-1.amazonaws.com
                           mountname=fsx
                           volumeType=filesystem
Events:                <none>
[hughliu@dev-dsk-hughliu-1e-2e6bb5a9 aws-fsx-openzfs-csi-driver]$ kubectl exec -it fsx-openzfs-app -- df -h
Filesystem                                             Size  Used Avail Use% Mounted on
overlay                                                 80G  5.4G   75G   7% /
tmpfs                                                   64M     0   64M   0% /dev
tmpfs                                                  3.8G     0  3.8G   0% /sys/fs/cgroup
fs-0edcf19e7e13f53ff.fsx.us-east-1.amazonaws.com:/fsx   14G  256K   14G   1% /data
/dev/nvme0n1p1                                          80G  5.4G   75G   7% /etc/hosts
shm                                                     64M     0   64M   0% /dev/shm
tmpfs                                                  6.9G   12K  6.9G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs                                                  3.8G     0  3.8G   0% /proc/acpi
tmpfs                                                  3.8G     0  3.8G   0% /sys/firmware
[hughliu@dev-dsk-hughliu-1e-2e6bb5a9 aws-fsx-openzfs-csi-driver]$ kubectl exec -it fsx-openzfs-app -- ls /data
out.txt  testVol
```

Successful mount of an FSx for OpenZFS child volume:
```
[hughliu@dev-dsk-hughliu-1e-2e6bb5a9 aws-fsx-openzfs-csi-driver]$ ka examples/static_provisioning/child_volumes/
pod/fsx-openzfs-app created
persistentvolume/fsx-openzfs-pv created
persistentvolumeclaim/fsx-openzfs-claim created
[hughliu@dev-dsk-hughliu-1e-2e6bb5a9 aws-fsx-openzfs-csi-driver]$ k describe pv fsx-openzfs-pv
Name:            fsx-openzfs-pv
Labels:          <none>
Annotations:     pv.kubernetes.io/bound-by-controller: yes
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    
Status:          Bound
Claim:           default/fsx-openzfs-claim
Reclaim Policy:  Delete
Access Modes:    RWX
VolumeMode:      Filesystem
Capacity:        50Gi
Node Affinity:   <none>
Message:         
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            fsx.openzfs.csi.aws.com
    FSType:            
    VolumeHandle:      fsvol-04bcd8f4403e24e05
    ReadOnly:          false
    VolumeAttributes:      dnsname=fs-0edcf19e7e13f53ff.fsx.us-east-1.amazonaws.com
                           mountname=/fsx/testVol
                           volumeType=volume
Events:                <none>
[hughliu@dev-dsk-hughliu-1e-2e6bb5a9 aws-fsx-openzfs-csi-driver]$ kubectl exec -it fsx-openzfs-app -- df -h
Filesystem                                                     Size  Used Avail Use% Mounted on
overlay                                                         80G  5.4G   75G   7% /
tmpfs                                                           64M     0   64M   0% /dev
tmpfs                                                          3.8G     0  3.8G   0% /sys/fs/cgroup
fs-0edcf19e7e13f53ff.fsx.us-east-1.amazonaws.com:/fsx/testVol   50G     0   50G   0% /data
/dev/nvme0n1p1                                                  80G  5.4G   75G   7% /etc/hosts
shm                                                             64M     0   64M   0% /dev/shm
tmpfs                                                          6.9G   12K  6.9G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs                                                          3.8G     0  3.8G   0% /proc/acpi
tmpfs                                                          3.8G     0  3.8G   0% /sys/firmware
[hughliu@dev-dsk-hughliu-1e-2e6bb5a9 aws-fsx-openzfs-csi-driver]$ kubectl exec -it fsx-openzfs-app -- ls /data
out.txt
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM --platform=$BUILDPLATFORM golang:1.19.0-bullseye as builder
```
Comment (msaggar) [inline Dockerfile:1]: why not use 1.20 for latest?

File: pkg/cloud/cloud.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+type cloud struct {
+	metadata MetadataService
+}
+
+// NewCloud returns a new instance of AWS cloud
+// It panics if session is invalid
+func NewCloud() (Cloud, error) {
+	sess := session.Must(session.NewSession(&aws.Config{}))
```
Comment (msaggar) [inline pkg/cloud/cloud.go:22]: kinda weird to return error but also have a panic, i would either change the name to `MustNewCloud` or `recover` and propagate the error if `session.Must` panics

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ALL_OS_ARCH_OSVERSION=$(foreach os, $(ALL_OS), ${ALL_OS_ARCH_OSVERSION_${os}})
+
+PLATFORM?=linux/amd64,linux/arm64
+
+# split words on hyphen, access by 1-index
+word-hyphen = $(word $2,$(subst -, ,$1))
+
+.EXPORT_ALL_VARIABLES:
+
+.PHONY: aws-openzfs-csi-driver
```
Comment (msaggar) [inline Makefile:38]: nit: `build-aws-openzfs-csi-driver`

File: hack/update-gomod [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
Comment (msaggar) [inline hack/update-gomod:1]: is this standard for csi drivers to manage deps like this?

File: pkg/driver/node.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+		return nil, status.Error(codes.InvalidArgument, "Volume capability missing in NodePublishVolumeRequest")
+	}
+	if !d.isValidVolumeCapabilities([]*csi.VolumeCapability{volCap}) {
+		return nil, status.Error(codes.InvalidArgument, "NodePublishVolume: Volume capability not supported")
+	}
+
+	context := req.GetVolumeContext()
+	dnsname := context["dnsname"]
+	mountname := context["mountname"]
+	volumeType := strings.ToLower(context["volumeType"])
```
Comment (msaggar) [inline pkg/driver/node.go:50]: if its a map lookup you might want to check if the value exists, even if there is some sort of static analysis where we would force this its good to validate server side

File: hack/update-gomod (hunk 4 above)
Comment (hughdanliu) [reply hack/update-gomod:1]: Yep - here are some examples from the NFS CSI driver: https://github.com/kubernetes-csi/csi-driver-nfs/blob/master/hack/update-gomod.sh
and the FSxL CSI driver: https://github.com/kubernetes-sigs/aws-fsx-csi-driver/blob/master/hack/update-gomod

File: deploy/kubernetes/base/csidriver.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+---
+
+apiVersion: storage.k8s.io/v1
+kind: CSIDriver
+metadata:
+  name: fsx.openzfs.csi.aws.com
+spec:
+  attachRequired: false
```
[the remark this answers is not in the document]
Comment (hughdanliu) [reply deploy/kubernetes/base/csidriver.yaml:8]: ack

File: Dockerfile (hunk 1 above)
Comment (hughdanliu) [reply Dockerfile:1]: Yeah, good point. This probably needs to be updated in all of the AWS CSI driver repos.

File: pkg/driver/node.go (hunk 5 above)
Comment (hughdanliu) [reply pkg/driver/node.go:50]: Go defaults non-existent key retrievals to the zero value of the type associated with the map (in this case an empty string). I have validation below these value retrievals to check for this, so I don't think I'd need to do another check here. I also have unit tests ensuring that the behavior is as expected when a user doesn't provide these values.

File: pkg/cloud/cloud.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+type Cloud interface {
+	GetMetadata() MetadataService
+}
+
+type cloud struct {
+	metadata MetadataService
```
[the remark this answers is not in the document]
Comment (hughdanliu) [reply pkg/cloud/cloud.go:16]: Yep - the implementation of this file is mostly a placeholder for until we implement dynamic provisioning. I'm partial to leaving this as is (and making you own making this change haha), but I can change this file if we feel that this is a blocker/bad practice. This came from copying from the file cache csi driver implementation process.

File: go.mod [hunk 8]
```diff
@@ -0,0 +1,76 @@
+module sigs.k8s.io/aws-fsx-openzfs-csi-driver
+
+go 1.19
```
Comment (jacobwolfaws) [inline go.mod:3]: go 1.19 here as well, probably should be go 1.20

File: deploy/kubernetes/base/csidriver.yaml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+---
+
+apiVersion: storage.k8s.io/v1
+kind: CSIDriver
+metadata:
+  name: fsx.openzfs.csi.aws.com
```
Comment (jacobwolfaws) [inline deploy/kubernetes/base/csidriver.yaml:6]: note: for file cache we have the metadata name as `filecache.csi.aws.com`, not sure if you have any thoughts on making it `fsxopenzfs.csi.aws.com` but wanted to bring it up

Comment (hughdanliu) [reply deploy/kubernetes/base/csidriver.yaml:6]: hmm I'd prefer to keep it as fsx.openzfs.csi.aws.com, it looks a lot better imo

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,13 @@
+FROM --platform=$BUILDPLATFORM golang:1.20.2-bullseye as builder
```
Comment (jacobwolfaws) [inline Dockerfile:1]: did you test the build on 1.20

File: go.mod [hunk 11]
```diff
@@ -0,0 +1,76 @@
+module sigs.k8s.io/aws-fsx-openzfs-csi-driver
+
+go 1.20
```
Comment (jacobwolfaws) [inline go.mod:3]: did you test mounting after updating this and building a new image

Comment (hughdanliu) [reply go.mod:3]: Yep

File: Dockerfile (hunk 10 above)
Comment (hughdanliu) [reply Dockerfile:1]: Yep

File: pkg/driver/node.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+	volCap := req.GetVolumeCapability()
+	if volCap == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume capability missing in NodePublishVolumeRequest")
+	}
+	if !d.isValidVolumeCapabilities([]*csi.VolumeCapability{volCap}) {
+		return nil, status.Error(codes.InvalidArgument, "NodePublishVolume: Volume capability not supported")
+	}
+
+	context := req.GetVolumeContext()
+	dnsname := context["dnsname"]
```
Comment (jacobwolfaws) [inline pkg/driver/node.go:48]: these should be stored as constants at the top of the file (mountname, dnsname, volumeType), I'm okay leaving this for the next commit though

**Comment by jacobwolfaws** [comment]
/approve
/lgtm

Nice!

**Pull request merged**